packages/core/src/modules/connections/DidExchangeProtocol.ts

```diff
@@ -0,0 +1,462 @@
+        multiUseInvitation: false,
+        routing,
+        autoAcceptConnection: connectionRecord.autoAcceptConnection,
+        invitation: connectionRecord.invitation,
+        tags: connectionRecord.getTags(),
+        protocol: connectionRecord.protocol,
+      })
+    }
+
+    const peerDid = DidPeer.fromDid(message.did)
+    let didDocument
+
+    if (peerDid.numAlgo === PeerDidNumAlgo.GenesisDoc) {
+      if (!message.didDoc) {
+        throw new AriesFrameworkError('No did doc')
+      }
+      await this.verifyAttachmentSignature(message.didDoc)
+      didDocument = JsonTransformer.fromJSON(message.didDoc.getDataAsJson(), DidDocument)
+    } else {
+      didDocument = peerDid.didDocument
+    }
+
+    const didRecord = new DidRecord({
+      id: message.did,
+      role: DidDocumentRole.Received,
+      // It is important to take the did document from the PeerDid class
+      // as it will have the id property
+      didDocument,
+      tags: {
+        // We need to save the recipientKeys, so we can find the associated did
+        // of a key when we receive a message from another connection.
+        recipientKeys: didDocument.recipientKeys,
+      },
+    })
+
+    await this.didRepository.save(didRecord)
+
+    connectionRecord.theirDid = message.did
+    connectionRecord.theirLabel = message.label
+    connectionRecord.threadId = message.id
+
+    await this.updateState(DidExchangeRequestMessage.type, connectionRecord)
+    this.logger.debug(`Process message ${DidExchangeRequestMessage.type} end`, connectionRecord)
+    return connectionRecord
+  }
+
+  public async createResponse(
+    connectionRecord: ConnectionRecord,
+    routing?: Routing
+  ): Promise<DidExchangeResponseMessage> {
+    this.logger.debug(`Create message ${DidExchangeResponseMessage.type} start`, connectionRecord)
+    DidExchangeStateMachine.assertCreateMessageState(DidExchangeResponseMessage.type, connectionRecord)
+
+    // They will then either update the provisional service information to rotate the key, or provision a new DID entirely.
+    // The choice here will depend on the nature of the DID used in the invitation.
+
+    // if reuse did from invitation then do ...
+    // otherwise create new did and didDoc
+    const { did, threadId } = connectionRecord
+
+    if (!threadId) {
+      throw new AriesFrameworkError('Missing threadId on connection record.')
+    }
+
+    // The responder will then craft an exchange response using the newly updated or provisioned information.
+
+    // Sign message attachment
+    // Use invitationKey by default, fall back to verkey (?)
+    const [verkey] = connectionRecord.invitation?.recipientKeys || []
+
+    if (!verkey) {
+      throw new AriesFrameworkError('Connection invitation does not contain recipient key.')
+    }
+
+    const peerDidRouting = routing || {
+      endpoints: connectionRecord.invitation?.serviceEndpoint ? [connectionRecord.invitation?.serviceEndpoint] : [],
```

Comment:
This should not be possible

Reply:
Do you mean that we should throw an error when there is no `routing` and no `connectionRecord.invitation`? Or something else?

Reply:
Yes, updated 👍